azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/manager/AccessibilityAnnouncementManager.kt

```diff
@@ -0,0 +1,110 @@
+                    val message = it.message(lastState, newState, rootView.context)
+                    if (message.isNotBlank()) {
+                        announce(activity, message)
+                    }
+                }
+            }
+            lastState = newState
+        }
+    }
+
+    private fun announce(activity: Activity, message: String) {
+        val manager = activity.getSystemService(Context.ACCESSIBILITY_SERVICE) as AccessibilityManager
+        if (manager.isEnabled) {
+            val event = AccessibilityEvent.obtain()
+            event.text.add(message)
+            event.eventType = AccessibilityEvent.TYPE_ANNOUNCEMENT
+            manager.sendAccessibilityEvent(event)
+        }
+    }
+}
+
+// Accessibility Hook
+//
+// shouldTrigger -> detect if we should announce something
+// message -> get the text tp read
+internal abstract class AccessibilityHook {
+    abstract fun shouldTrigger(lastState: ReduxState, newState: ReduxState): Boolean
+    abstract fun message(lastState: ReduxState, newState: ReduxState, context: Context): String
+}
+
+// Hook to announce when participants join/leave a meeting
+internal class ParticipantAddedOrRemovedHook : AccessibilityHook() {
+    var callJoinTime = System.currentTimeMillis()
+    override fun shouldTrigger(lastState: ReduxState, newState: ReduxState): Boolean {
+        if (lastState.callState.callingStatus != CallingStatus.CONNECTED && newState.callState.callingStatus == CallingStatus.CONNECTED) {
+            callJoinTime = System.currentTimeMillis()
+            return false
+        }
+        val shouldRun = lastState.remoteParticipantState.participantMap.size != newState.remoteParticipantState.participantMap.size
+
+        if (shouldRun && (System.currentTimeMillis() - callJoinTime) > 1000) {
+            return true
+        }
+
+        return false
+    }
+
+    override fun message(lastState: ReduxState, newState: ReduxState, context: Context): String {
+        val oldList = lastState.remoteParticipantState.participantMap.values
+        val newList = newState.remoteParticipantState.participantMap.values
+
+        val added = newList.filter { !oldList.contains(it) }
+        val removed = oldList.filter { !newList.contains(it) }
+
+        var result = ""
+        if (added.size == 1) {
+            result = context.getString(R.string.azure_communication_ui_accessibility_user_added, added.first().displayName)
+        } else if (removed.size == 1) {
+            result = context.getString(R.string.azure_communication_ui_accessibility_user_left, removed.first().displayName)
+        }
+        return result
+    }
+}
+
+// Hook to announce the meeting was successfully joined
+internal class MeetingJoinedHook : AccessibilityHook() {
+    override fun shouldTrigger(lastState: ReduxState, newState: ReduxState): Boolean {
+        return (lastState.callState.callingStatus != CallingStatus.CONNECTED && newState.callState.callingStatus == CallingStatus.CONNECTED)
```

Comment:
can we change this to:
`lastState.callState.callingStatus == CallingStatus.CONNECTING && newState.callState.callingStatus == CallingStatus.CONNECTED`

Reply:
It was chosen this way specifically to be "now in connected state". The previous state may not be "connecting".

E.g. CALL_STATE has
    NONE,
    EARLY_MEDIA,
    CONNECTING,
    RINGING,
    CONNECTED,
    LOCAL_HOLD,
    DISCONNECTING,
    DISCONNECTED,
    IN_LOBBY,
    REMOTE_HOLD;
    
    
    It could be IN_LOBBY->CONNECTED, or EARLY_MEDIA->CONNECTED etc.
    
    I specifically chose last != connected and current == connected so that it captures all possible transitions into connected state.

Reply:
👍